src/cargo/ops/cargo_compile.rs

```diff
@@ -857,7 +864,27 @@ fn generate_targets(
             let features_for = FeaturesFor::from_for_host(target.proc_macro());
             let features = resolved_features.activated_features(pkg.package_id(), features_for);
 
-            for kind in requested_kinds {
+            // If `--target` has not been specified, then the unit
+            // graph is built almost like if `--target $HOST` was
+            // specified. See `rebuild_unit_graph_shared` for more on
+            // why this is done. However, if the package has its own
+            // `package.target` key, then this gets used instead of
+            // `$HOST`
+            let explicit_kinds = if let Some(k) = pkg.manifest().forced_kind() {
+                vec![k]
+            } else {
+                requested_kinds
+                    .iter()
+                    .map(|kind| match kind {
+                        CompileKind::Host => {
+                            pkg.manifest().default_kind().unwrap_or(explicit_host_kind)
+                        }
+                        CompileKind::Target(t) => CompileKind::Target(*t),
+                    })
+                    .collect()
+            };
```

Comment:
I think that this may not actually be the right place for this conditional (or it may need to be in more than one location). For example if a build script depends on a wasm blob (for whatever reason) I don't think that's factored in here? 

I'm not sure if it's possible to easily expose this today really, but when we can depend on binary artifacts themselves (binaries, cdylibs, etc), we'll need to factor in manifest-requested targets to the `Unit` graph construction steps.

Reply:
I'm not sure I understand correctly here. Do you mean that, when binary artifact dependencies happen in `cargo`, then this code will need to be overhauled again to properly handle this case? 

Or is it already possible for a build script to depend on an artifact generated by another target, which would mean that there is an action to be undertaken now?

Reply:
No I think you're right, without artifact dependencies this doesn't really matter. I think this is probably good enough for now on a second pass. This'll just need to be updated when we get artifact dependencies.